packages/db/postgres/functions.sql

```diff
@@ -250,6 +250,95 @@ BEGIN
 END
 $$;
 
+DO 
+$$
+BEGIN
+  IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'used_storage') THEN
+    CREATE TYPE used_storage AS (uploaded TEXT, pinned TEXT, total TEXT);
+  END IF;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION user_used_storage(query_user_id BIGINT) 
+RETURNS used_storage
+LANGUAGE plpgsql
+AS
+$$
+DECLARE 
+  used_storage  used_storage;
+  uploaded      BIGINT;
+  pinned        BIGINT;
+  total         BIGINT;
+BEGIN
+  uploaded := 
+    (
+      SELECT COALESCE(SUM(c.dag_size), 0)
+      FROM upload u
+      JOIN content c ON c.cid = u.content_cid
+      WHERE u.user_id = query_user_id::BIGINT
+      AND u.deleted_at is null
+    );
+
+  pinned := 
+    (
+      SELECT COALESCE((
+        SELECT COALESCE(SUM(c.dag_size), 0)
+        FROM psa_pin_request psa_pr
+        JOIN content c ON c.cid = psa_pr.content_cid
+        JOIN pin p ON p.content_cid = psa_pr.content_cid
+        JOIN auth_key a ON a.id = psa_pr.auth_key_id
+        WHERE a.user_id = query_user_id::BIGINT
+        AND psa_pr.deleted_at is null
+        AND p.status = 'Pinned'
+        GROUP BY p.status
+      ), 0)
+    ); 
+
+  total := uploaded + pinned;
+
+  SELECT  uploaded::TEXT,
+          pinned::TEXT,
+          total::TEXT
+  INTO    used_storage;
+
+  return used_storage;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION users_by_storage_used(
+  from_percent INTEGER,
+  to_percent INTEGER DEFAULT NULL
+)
+  RETURNS TABLE
+    (
+      id            TEXT,
+      name          TEXT,
+      email         TEXT,
+      storage_quota TEXT,
+      storage_used  TEXT
+    )
+LANGUAGE plpgsql
+AS
+$$
+DECLARE
+  -- Default storage quota 1TiB = 1099511627776 bytes
+  default_quota BIGINT := 1099511627776;
+BEGIN
+  RETURN QUERY
+    SELECT  u.id::TEXT                                      AS id,
+            u.name                                          AS name,
+            u.email                                         AS email,
+            COALESCE(ut.value::BIGINT, default_quota)::TEXT AS storage_quota,
+            (user_used_storage(u.id)).total                 AS storage_used
+    FROM public.user u
+    FULL OUTER JOIN user_tag ut ON u.id = ut.user_id
+    WHERE (ut.tag IS NULL OR (ut.tag = 'StorageLimitBytes' AND ut.deleted_at IS NULL))
+    AND (user_used_storage(u.id)).total::BIGINT >= (from_percent::NUMERIC/100) * COALESCE(ut.value::BIGINT, default_quota)
```

Comment:
I think with the query written this way, `user_used_storage` stored procedure might be run multiple times for the same user. We should make sure it is ran once per user.

Reply:
It could be something like this 
```
RETURN QUERY
    SELECT *
    FROM (
        SELECT  u.id::TEXT                                AS id,
          u.name                                          AS name,
          u.email                                         AS email,
          COALESCE(ut.value::BIGINT, default_quota)::BIGINT AS storage_quota,
          (user_used_storage(u.id)).total::BIGINT                   AS storage_used
      FROM public.user u
      LEFT JOIN user_tag ut ON u.id = ut.user_id
      WHERE (ut.tag IS NULL OR (ut.tag = 'StorageLimitBytes' AND ut.deleted_at IS NULL))
    ) sq
    WHERE sq.storage_used >= (from_percent::NUMERIC/100) * sq.storage_quota
    AND (to_percent IS NULL OR sq.storage_used < (to_percent::NUMERIC/100) * sq.storage_quota)
    ORDER BY sq.storage_used DESC;
END
```

Reply:
Thanks this is super helpful. Have implemented.